File: Range_Sum_Of_BST/js.d/tree_range_of_sum.js
// [10,5,15,3,7,13,18,1,null,6] => becomes
// L=7, R=15 => 45
/*
                   10
                 /    \
                5     15
              /   \  /   \
             3    7 13   18
           /     /
          1     6
*/
// LeetCode definition for a binary tree node.
/*
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}
*/

// ES6 implementation
class TreeNode {
  constructor (val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
  }
  // Node insertion function
  insertNode (val, location){
    //location = (location===undefined ? null : location)
    if (val){
      if (!location){
        console.log(`Inserting node value == ${val}\n`)
        location = new TreeNode(val);
      } else {
        if (val < location.val){
          // go left
          if (! location.left){
            location.left = new TreeNode(val);
          } else {
            console.log(`${val} < ${location.val}: going left`)
            this.insertNode(val, location.left)
          }
        } else {
          // go right
          if (! location.right){
            location.right = new TreeNode(val);
          } else {
            console.log(`${val} > ${location.val}: going right`)
            this.insertNode(val, location.right)
          }
        }
      }
    }
  }
};

/** LeetCode function - passing in constructed BST
 * @param {TreeNode} root
 * @param {number} L
 * @param {number} R
 * @return {number}
 */
var rangeSumBST = function(root, L, R) {
  let stack = [root];
  let rangeNodes = [];
  let findNodesInRange = (node) => {
    if (node && node.val >= L && node.val <= R) rangeNodes.push(node.val);
    if (node.left && node.val >= node.left.val) stack.push(node.left);
    if (node.right && node.val <= node.right.val) stack.push(node.right);
  };
  while (stack.length > 0){
    findNodesInRange(stack.shift());
  }
  return rangeNodes.reduce((a,b) => a+b, 0);
};


let nodeVals = [10,5,15,3,7,13,18,1,null,6];
var tn = new TreeNode(nodeVals.shift())
nodeVals.map(n => tn.insertNode(n, tn));

console.log(rangeSumBST(tn, 7, 15));

File: Range_Sum_Of_BST/py.d/tree_range_of_sum.py
from typing import List
from collections import deque

# [10,5,15,3,7,13,18,1,null,6] => becomes
# L=7, R=15 => 45
'''
                   10
                 /    \
                5     15
              /   \  /   \
             3    7 13   18
           /     /
          1     6
'''
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = None
    self.right = None

  def insert_node(self, val, location=None):
    '''
    Working constructor function for LeetCode b-tree questions
    (where b-tree is stated as list of ints and null values)
    '''
    if not location:
      location = TreeNode(val)
    else:
      if val > location.val:
        # we go right
        if not location.right:
          location.right = TreeNode(val)
        else:
          self.insert_node(val, location=location.right)
      elif val < location.val:
        # we go left
        if not location.left:
          location.left = TreeNode(val)
        else:
          self.insert_node(val, location=location.left)

  @classmethod
  def ascend_tree(cls, node=None):
    if node:
      cls.ascend_tree(node.left)
      print(node.val)
      cls.ascend_tree(node.right)

class Solution:
  def __init__(self):
    self.range_nodes = []

  def seek_nodes_in_range(self, L, R, node=None):
    if node:
      if node.val > L: self.seek_nodes_in_range(L, R, node.left)
      if node.val >= L and node.val <= R:
        #print(f'Node {node.val} would be appended')
        self.range_nodes.append(node.val)
      if node.val < R: self.seek_nodes_in_range(L, R, node.right)

  def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
    self.seek_nodes_in_range(L, R, root)
    return sum(self.range_nodes)

if __name__=="__main__":
  sol = Solution()
  # [10,5,15,3,7,13,18,1,None,6], L=7, R=15 => 45
  tree_vals = [10,5,15,3,7,13,18,1,None,6]
  left, right = (7, 15)

  tree_vals = deque([val for val in tree_vals if type(val) == int])
  new_tree = TreeNode(tree_vals.popleft())

  # insert all nodes into new_tree
  while tree_vals:
    new_tree.insert_node(tree_vals.popleft(), location=new_tree)

  # ascend tree and display node values
  TreeNode.ascend_tree(node=new_tree) # => works

  print(sol.rangeSumBST(new_tree, left, right)) # should return 45
